### ISSUE TITLE
Java 15 support

### ISSUE THREAD

**Issue by ollierob** [google/guice#1409]
Please add JDK 15 support. Version 4.2.3 currently doesn't support it, for example:
```
java.lang.IllegalArgumentException: Unsupported class file major version 59
	at com.google.inject.internal.asm.$ClassReader.<init>(ClassReader.java:195)
	at com.google.inject.internal.asm.$ClassReader.<init>(ClassReader.java:176)
	at com.google.inject.internal.asm.$ClassReader.<init>(ClassReader.java:162)
	at com.google.inject.internal.asm.$ClassReader.<init>(ClassReader.java:283)
	at com.google.inject.internal.util.LineNumbers.<init>(LineNumbers.java:69)
	at com.google.inject.internal.util.StackTraceElements$1.load(StackTraceElements.java:49)
	at com.google.inject.internal.util.StackTraceElements$1.load(StackTraceElements.java:45)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
	... 52 common frames omitted
```

**Comment by GedMarc** [comment]
you can swap out for guicedee which also contains the fix for the encapsulation warning

https://search.maven.org/artifact/com.guicedee.services/guice
```
<dependency>
  <groupId>com.guicedee.services</groupId>
  <artifactId>guice</artifactId>
  <version>1.0.19.2-jre15</version>
</dependency>
```

All extensions as well in the same maven group-id 
https://www.guicedee.com

**Comment by tylerscoville** [comment]
I see https://github.com/google/guice/pull/1400 was opened. 

Is there any timeline on when guice will support java 15?

**Comment by electrum** [comment]
Is there anything we can do to move this along? Guice is our current blocker for moving to Java 15.

**Comment by ganeshbhargav** [comment]
`Error: java.lang.IllegalArgumentException: Unsupported class file major version 59`

Explanation: Class file Major version 59 corresponds to Java 15. You are using Java 15 (and compiling your code for it so the generated class files are only valid for Java 15 and newer).


The easy fix is to reconfigure your build to compile for Java 14 or earlier depending on what version of Java you need to use.

You can set the version of compiler at compile time. And compile your java code into old versions of java.

From Oracle article : http://docs.oracle.com/javase/1.5.0/docs/tooldocs/windows/javac.html

Cross-Compilation Example

Here we use javac to compile code that will run on a 1.4 VM.

Command: 
```
javac -target 1.4 -bootclasspath jdk1.4.2/lib/classes.zip \
             -extdirs "" OldCode.java
```

You might also need following parameter to set denote the version of your code.

`-source release`
-Specifies the version of source code accepted.

This should support Apps developed with Java version 15 and newer and not able to run guice.

Hope this helps resolve the issue.

**Comment by markmarch** [comment]
We've just released 5.0.0 beta, with Java 15 support, please test it with your project and report bugs.

See https://github.com/google/guice/wiki/Guice500
